crates/ruff/src/source_kind.rs

```diff
@@ -15,4 +16,23 @@ impl SourceKind {
             None
         }
     }
+
+    #[must_use]
+    pub(crate) fn updated(&self, new_source: String, source_map: &SourceMap) -> Self {
+        match self {
+            SourceKind::Jupyter(notebook) => {
+                let mut cloned = notebook.clone();
```

Comment:
This for sure is more expensive than the old "update in place".

Reply:
Yeah, my main reason to update in place was to avoid cloning the entire notebook. Actually, the only thing we need to update with each linting cycle are the cell markers and the transformed source code can be looped over to the next iteration cycle. The cell content can be updated at the end before we actually update the Notebook that too only when `--fix` is passed.

Using the above idea, we could create a `NotebookView` which would only contain the necessary fields (for now I think it's cell markers only and the concatenated source code?) and should be much cheaper to clone. We would not even clone it and use an `updated` method on the view with the `SourceMap` and transformed source code and create a new view with the updated source code and cell markers. Once the linter iteration is complete, we can _apply_ the view onto the Notebook, update the cell content then and write the Notebook back to the disk.

I think the abstraction would be to have a view into the source code be it Python or Jupyter and the view would have `updated` method to update the view. At the end the view will be applied to the concrete type and saved onto the disk.

This is just a rough idea based on my morning walk :)

Reply:
I don't think I'm able to follow. Do you want to take a stab at this together with performing the renaming that you suggested?

I tried to change `raw` to an `Rc` to get cheap cloning. However, it turns out that `update_cell_content` mutates the raw cells. The only larger structure that doesn't seem to get updated is `valid_code_cells`. 


I think we could also move out `index` from the `Notebook` because it is only used in `Diagnostics` and instead have a `JupyterIndex` that is either `Lazy(Notebook)` or `Built(Index)`.

Reply:
> Do you want to take a stab at this together with performing the renaming that you suggested?

Yeah, I can give it a go next week.

> Or is this something that needs changing before landing this PR?

No

> The only larger structure that doesn't seem to get updated is `valid_code_cells`.

What do you mean here by "larger structure"? As `valid_code_cells` is just a vector of `u32`.


> However, it turns out that `update_cell_content` mutates the raw cells.

Yeah, this update isn't really required for every linter loop. We can avoid doing that and only update it at the end. This basically uses the concatenated source code and cell markers to update the cell content.

> I think we could also move out `index` from the `Notebook` because it is only used in `Diagnostics` and instead have a `JupyterIndex` that is either `Lazy(Notebook)` or `Built(Index)`.

Yeah, this is a good idea.

Reply:
> What do you mean here by "larger structure"? As valid_code_cells is just a vector of u32.

With large, I mean any non-fixed size heap allocated data structure that requires a deep clone.